### DISCUSSION TITLE
Can AutoGOAL be used for program synthesis?

### DISCUSSION THREAD

**Discussion by pauljurczak** [autogoal/autogoal#172] [Q&A]
Can AutoGOAL be used for program synthesis? For example, find a Python function `f`, which satisfies these assertions:
```
assert(f(3, 1) == 3)
assert(f(2, 2) == 2)
assert(f(1, 3) == 1)
assert(f(4, 4) == 4)
assert(f(8, 5) == 9)
assert(f(7, 6) == 9)
assert(f(9, 7) == 9)
assert(f(5, 8) == 9)
assert(f(6, 9) == 9)
```
I expect a result equivalent to:
```
def f(a, b):
  return a if b <= 4 else 9
```
It is acceptable that synthetic `f` is correct only for small integral arguments, e.g. `in range(10)`. If needed, constraints on function syntax can be imposed.

**Comment by apiad** [comment]
Hi! Yes, it should be usable, but we don't currently have a predefined search space of, e.g., arbitrary Python programs. So the user would need to provide a grammar specifically for this domain. This is something we want to look into but we don't have the bandwidth now to go full unrestricted program synthesis, so any help or suggestion is really appreciated :)
